### PR TITLE
Send RemoteNewTabLocation values via message

### PR DESCRIPTION
**Pull request by oyiptong**
r? @marcoscaceres or @emtwo

### CODE REVIEW

**Comment by oyiptong** [comment]
or r? @sarracini

File: browser/base/content/remote-newtab/newTab.js [hunk 1]
```diff
@@ -40,11 +39,10 @@
     return commandHandled;
   }
 
-  function initRemotePage() {
+  function initRemotePage(initData) {
     // Messages that the iframe sends the browser will be passed onto
     // the privileged parent process
-    let iframe = getIframe();
-    iframe.src = imports.RemoteNewTabLocation.href;
+    remoteNewTabLocation = initData;
```
Comment (marcoscaceres) [inline browser/base/content/remote-newtab/newTab.js:44]: Probably don't bother aliasing this?

Comment (marcoscaceres) [inline browser/base/content/remote-newtab/newTab.js:44]: Maybe just rename s/initData/remoteNewTabLocation/

File: browser/base/content/remote-newtab/newTab.js [hunk 2]
```diff
@@ -102,6 +104,10 @@
     return iframe;
   }
 
-  // Everything is loaded. Initialize the New Tab Page.
-  initRemotePage();
+  addMessageListener("NewTabFrame:init", function foo(message) {
```
Comment (marcoscaceres) [inline browser/base/content/remote-newtab/newTab.js]: heh, "foo" is my fault here :) Maybe rename it "handler".

File: browser/base/content/remote-newtab/newTab.js [hunk 3]
```diff
@@ -102,6 +104,10 @@
...
+    // Everything is loaded. Initialize the New Tab Page.
+    removeMessageListener("NewTabFrame:init", foo);
```
Comment (marcoscaceres) [inline browser/base/content/remote-newtab/newTab.js]: same here, s/`foo`/`handler`

File: browser/components/newtab/RemoteAboutNewTab.jsm [hunk 4]
```diff
@@ -49,6 +51,12 @@ let RemoteAboutNewTab = {
     this.pageListener.addMessageListener("NewTab:ReportSitesAction", this.reportSitesAction.bind(this));
     this.pageListener.addMessageListener("NewTab:SpeculativeConnect", this.speculativeConnect.bind(this));
     this.pageListener.addMessageListener("NewTab:RecordSiteClicked", this.recordSiteClicked.bind(this));
+    this.pageListener.addMessageListener("NewTabFrame:GetInit", () => {
+      this.pageListener.sendAsyncMessage("NewTabFrame:init", {
```
Comment (marcoscaceres) [inline browser/components/newtab/RemoteAboutNewTab.jsm]: s/NewTabFrame:init/NewTabFrame:Init (capital I on Init)

File: browser/base/content/remote-newtab/newTab.js (hunk 2 above)
Comment (marcoscaceres) [inline browser/base/content/remote-newtab/newTab.js]: s/NewTabFrame:init/NewTabFrame:Init/g

File: browser/base/content/remote-newtab/newTab.js (hunk 1 above)
Comment (oyiptong) [inline browser/base/content/remote-newtab/newTab.js:44]: It's used outside of this scope

**Comment by marcoscaceres** [comment]
Couple of small nits. But r+ otherwise.

File: browser/components/newtab/RemoteAboutNewTab.jsm [hunk 5]
```diff
@@ -446,10 +456,10 @@ let RemoteAboutNewTab = {
           RemoteNewTabUtils.blockedLinks.resetCache();
           break;
       }
-    } else if (aTopic == "browser:purge-session-history") {
-        RemoteNewTabUtils.links.resetCache();
-        RemoteNewTabUtils.links.populateCache(() => {
-          this.pageListener.sendAsyncMessage("NewTab:UpdateLinks", {
+    } else if (aTopic === "browser:purge-session-history") {
```
Comment (marcoscaceres) [inline browser/components/newtab/RemoteAboutNewTab.jsm:493]: Nit: this `else if` is a bit ugly (particularly as it's not followed by a `} else {` ). Can it just be an normal if on its own on a new line?

File: browser/components/newtab/tests/browser/dummy_page.html [hunk 6]
```diff
@@ -0,0 +1,19 @@
+<!DOCTYPE html>
+
+<html>
```
Comment (marcoscaceres) [inline browser/components/newtab/tests/browser/dummy_page.html:3]: please add after `<html>`: 

```
<meta charset="utf-8">
```

Stops the annoying warning about lack of content type in the console.

File: browser/components/newtab/tests/browser/browser_remotenewtab_pageloads.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+  RemoteNewTabLocation.override(TEST_URL);
+
+  let frameLoadPromise = new Promise(resolve => {
+    RemoteAboutNewTab.pageListener.addMessageListener("NewTab:testMessage", () => {
+      ok(true, "message received");
+      resolve();
+    });
+  });
+
+  let loaded = new Promise(resolve => {
```
Comment (marcoscaceres) [inline browser/components/newtab/tests/browser/browser_remotenewtab_pageloads.js]: You might want instead:

``` JS
add_task(function* () {
    let tabOptions = {
      gBrowser,
      url: "about:remote-newtab",
    };
    yield BrowserTestUtils.withNewTab(
      tabOptions,
      browser => thingYouWantToTest
    );

   function* thingYouWantToTest(){

   }
});
```

File: browser/components/newtab/RemoteAboutNewTab.jsm [hunk 8]
```diff
@@ -424,10 +434,10 @@ let RemoteAboutNewTab = {
    * also clear the links cache and update the set of links to display, as they
    * may have changed, then proceed with the page update.
    */
-  observe: function(aSubject, aTopic, aData) {
+  observe: function(aSubject, aTopic, aData) { // jshint ignore:line
```
Comment (marcoscaceres) [inline browser/components/newtab/RemoteAboutNewTab.jsm:471]: Nit: We've been kinda ignoring Hungarian notation in this JSM (i.e., for all the `message` args)... I wonder if we should do the same here?

**Comment by marcoscaceres** [comment]
R+. Just a couple of nits, and suggestion about alternative way of doing the testing.

File: browser/components/newtab/RemoteAboutNewTab.jsm (hunk 5 above)
Comment (oyiptong) [inline browser/components/newtab/RemoteAboutNewTab.jsm:493]: sorry, not code i wrote in this patch. I just fixed linting problems

Comment (marcoscaceres) [inline browser/components/newtab/RemoteAboutNewTab.jsm:493]: np

File: browser/components/newtab/RemoteAboutNewTab.jsm (hunk 8 above)
Comment (oyiptong) [inline browser/components/newtab/RemoteAboutNewTab.jsm:471]: as above, sorry, not code i wrote in this patch. I just fixed linting problems

Comment (marcoscaceres) [inline browser/components/newtab/RemoteAboutNewTab.jsm:471]: np

**Pull request closed without merging**